### PR TITLE
Remove allocations in callbacks 

### PR DESCRIPTION
**Pull request by frapac**
- add `PrimalVector` abstraction to remove the need to use `_madnlp_unsafe_wrapper` 
- rework KKT system to pass only the Jacobian w.r.t. the original variable `x` (without the slack) 
-

### CODE REVIEW

File: src/IPM/callbacks.jl [hunk 1]
```diff
@@ -1,128 +1,138 @@
...
     nlp = solver.nlp
     cnt = solver.cnt
     @trace(solver.logger,"Evaluating Lagrangian Hessian.")
-    dual(solver._w1) .= l.*solver.con_scale
+    dual(solver._w1) .= l .* solver.con_scale
     hess = get_hessian(kkt)
-    x_nlpmodel = _madnlp_unsafe_wrap(x, get_nvar(nlp))
-    hess_nlpmodel = _madnlp_unsafe_wrap(hess, get_nnzh(nlp.meta))
+    scale = (get_minimize(nlp) ? one(T) : -one(T))
+    scale *= (is_resto ? zero(T) : solver.obj_scale[])
```
Comment (sshin23) [inline src/IPM/callbacks.jl:80]: `scale = is_resto ? zero(T) : get_minimize(nlp) ? one(T) : -one(T)`

File: test/madnlp_test.jl [hunk 2]
```diff
@@ -118,6 +118,30 @@ end
...
+    @test n_allocs == 16 # objective is still allocating
+    n_allocs = @allocated MadNLP.eval_grad_f_wrapper!(solver, f, x)
+    @test n_allocs == 0
+    n_allocs = @allocated MadNLP.eval_cons_wrapper!(solver, c, x)
+    @test n_allocs == 0
+    n_allocs = @allocated MadNLP.eval_jac_wrapper!(solver, kkt, x)
+    @test n_allocs == 0
+    n_allocs = @allocated MadNLP.eval_lag_hess_wrapper!(solver, kkt, x, solver.y)
+    @test n_allocs == 0
+end
```
Comment (sshin23) [inline test/madnlp_test.jl:143]: this is nice! maybe in the future we can do it for the whole solver level

File: src/KKT/sparse.jl [hunk 3]
```diff
@@ -7,6 +7,7 @@ Implement the [`AbstractReducedKKTSystem`](@ref) in sparse COO format.
 """
 struct SparseKKTSystem{T, VT, MT} <: AbstractReducedKKTSystem{T, VT, MT}
     hess::VT
+    jac_callback::VT
```
Comment (sshin23) [inline src/KKT/sparse.jl:10]: what does this do?

**Review by sshin23 — APPROVED** [review]
Nice work, @frapac! It overall looks great! I only have a few comments

File: test/madnlp_test.jl (hunk 2 above)
Comment (frapac) [reply test/madnlp_test.jl:143]: I agree! we should at least apply the same test on the linear solver wrapper.

**Pull request merged**